File: randoom/Virat_s_Century_and_India_s_Win.java
package randoom;
import java.util.Scanner;

public class Virat_s_Century_and_India_s_Win {

    private static final int MOD = 1000000007;

    public static double calculateProbability(int N, int b, int w, int x) {
        double[][][] dp = new double[N + 1][b + 1][w + 1];

        for (int r = 0; r <= N; r++) {
            for (int balls = 0; balls <= b; balls++) {
                for (int wickets = 0; wickets <= w; wickets++) {
                    if (r == 0 && wickets <= 10) {
                        dp[r][balls][wickets] = (x >= 100) ? 1.0 : 0.0;
                    } else if (balls == 0 || wickets == 10) {
                        dp[r][balls][wickets] = 0.0;
                    } else {
                        double probability = 0.0;
                        for (int run : new int[]{0, 1, 2, 3, 4, 6}) {
                            int newRemainingRuns = r - run;
                            int newBallsLeft = balls - 1;
                            int newWicketsDown = (run > 0) ? wickets : wickets + 1;
                            int newViratScore = x + run;

                            // Update state based on run
                            if (run % 2 == 1 && run > 0) {
                                // Switch ends
                                newViratScore += run;
                            } else if (run % 2 == 1 || (balls == 1 && wickets < 10)) {
                                // Switch ends on odd runs or last ball of the over (if wickets are less than 10)
                                newBallsLeft -= 1;
                            }

                            if (newRemainingRuns >= 0 && newBallsLeft >= 0 && newWicketsDown <= 10) {
                                probability += (1.0 / 6) * dp[newRemainingRuns][newBallsLeft][newWicketsDown];
                            }
                        }
                        dp[r][balls][wickets] = probability % MOD;
                    }
                }
            }
        }

        // Calculate p * q^-1 modulo 10^9 + 7
        int p = (int) (dp[N][b][w] * modInverse(b, MOD)) % MOD;

        return p;
    }

    private static int modInverse(int a, int m) {
        int m0 = m;
        int y = 0, x = 1;

        while (a > 1) {
            int q = a / m;
            int t = m;

            m = a % m;
            a = t;
            t = y;

            y = x - q * y;
            x = t;
        }

        // Make sure x is positive
        return (x % m0 + m0) % m0;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt(); // Number of test cases

        for (int i = 0; i < t; i++) {
            int N = sc.nextInt(); // Required runs
            int b = sc.nextInt(); // Balls left
            int w = sc.nextInt(); // Wickets down
            int x = sc.nextInt(); // Virat's current score

            double initialProbability = calculateProbability(N, b, w, x);

            System.out.println(initialProbability);
        }
    }
}

File: randoom/Virat_s_Century_and_India_s_Win.cpp
#include <iostream>
#include <vector>
using namespace std;

const int MOD = 1e9 + 7;

vector<vector<vector<vector<int>>>> memo(151, vector<vector<vector<int>>>(61, vector<vector<int>>(11, vector<int>(101, -1))));

int dp(int i, int j, int k, int l, int N) {
    if (i == 0 && j == 0 && k == 10) {
        return (l >= 100) ? 1 : 0;
    }
    if (j == 0 || k == 10) {
        return 0;
    }

    if (memo[i][j][k][l] != -1) {
        return memo[i][j][k][l];
    }

    int total_ways = 0;
    total_ways = (total_ways + dp(i - l, j - 1, k + 1, l, N)) % MOD;       // Wicket
    total_ways = (total_ways + dp(i - l + 1, j - 1, k + 1, l + 1, N)) % MOD; // 1 run
    total_ways = (total_ways + dp(i - l + 2, j - 1, k + 1, l + 2, N)) % MOD; // 2 runs
    total_ways = (total_ways + dp(i - l + 3, j - 1, k + 1, l + 3, N)) % MOD; // 3 runs
    total_ways = (total_ways + dp(i - l + 4, j - 1, k + 1, l + 4, N)) % MOD; // 4 runs
    total_ways = (total_ways + dp(i - l + 6, j - 1, k + 1, l + 6, N)) % MOD; // 6 runs

    memo[i][j][k][l] = total_ways;
    return total_ways;
}

int solve(int N, int b, int w, int x) {
    int total_ways = dp(N, b, w, x, N);
    total_ways %= MOD;

    // Calculate the probability p * q^-1 modulo 10^9 + 7
    long long inverse_q = 1;
    for (int i = 0; i < 6; ++i) {
        inverse_q = (inverse_q * 6) % MOD;
    }

    long long result = (total_ways * inverse_q) % MOD;

    return result;
}

int main() {
    int T;
    cin >> T;

    for (int t = 0; t < T; ++t) {
        int N, b, w, x;
        cin >> N >> b >> w >> x;
        int result = solve(N, b, w, x);
        cout << result << endl;
    }

    return 0;
}